include/irr/asset/filters/CSummedAreaTableImageFilter.h

```diff
@@ -0,0 +1,382 @@
+				core::smart_refctd_ptr<asset::ICPUBuffer>	columnSumCache;					//!< pointer of column extent size for x values used for holding max values going down till 0 is achievied
+				constexpr E_SUM_MODE mode = ExclusiveMode ? ESM_EXCLUSIVE : ESM_INCLUSIVE;
+		};
+
+	protected:
+		CSummedAreaTableImageFilterBase() {}
+		virtual ~CSummedAreaTableImageFilterBase() {}
+
+		static inline uint32_t getRequiredScratchByteSize(E_FORMAT outFormat, const core::vectorSIMDu32& outExtent = core::vectorSIMDu32(0,0,0,0))
+		{
+			return asset::getTexelOrBlockBytesize(outFormat) * outExtent.X * outExtent.Y * outExtent.Z;
+		}
+
+		static inline bool validate(CStateBase* state)
+		{
+			if (!state)
+				return false;
+
+			return true;
+		}
+};
+
+//! Fill texel buffer with computed sum of left and down texels placed in input image
+/*
+	When the summing is in exclusive mode - it computes the sum of all the pixels placed
+	on the left and down for a new single texel but it doesn't take sum the main texel itself.
+	In inclusive mode, the texel we start from is taken as well and added to the sum.
+*/
+
+template<bool ExclusiveMode = false>
+class CSummedAreaTableImageFilter : public CImageFilter<CSummedAreaTableImageFilter>, public CSummedAreaTableImageFilterBase<ExclusiveMode>
+{
+	public:
+		virtual ~CSummedAreaTableImageFilter() {}
+
+		class CPrototypeState : public IImageFilter::IState
+		{
+			public:
+				CPrototypeState()
+				{
+					inOffsetBaseLayer = core::vectorSIMDu32();
+					inExtentLayerCount = core::vectorSIMDu32();
+					outOffsetBaseLayer = core::vectorSIMDu32();
+					outExtentLayerCount = core::vectorSIMDu32();
+				}
+				CPrototypeState(const CPrototypeState& other) : inMipLevel(other.inMipLevel), outMipLevel(other.outMipLevel), inImage(other.inImage), outImage(other.outImage)
+				{
+					inOffsetBaseLayer = other.inOffsetBaseLayer;
+					inExtentLayerCount = other.inExtentLayerCount;
+					outOffsetBaseLayer = other.outOffsetBaseLayer;
+					outExtentLayerCount = other.outExtentLayerCount;
+				}
+				virtual ~CPrototypeState() {}
+
+				union
+				{
+					core::vectorSIMDu32	inOffsetBaseLayer;
+					struct
+					{
+						VkOffset3D		inOffset;
+						uint32_t		inBaseLayer;
+					};
+				};
+				union
+				{
+					core::vectorSIMDu32 inExtentLayerCount;
+					struct
+					{
+						VkExtent3D		inExtent;
+						uint32_t		inLayerCount;
+					};
+				};
+				union
+				{
+					core::vectorSIMDu32 outOffsetBaseLayer;
+					struct
+					{
+						VkOffset3D		outOffset;
+						uint32_t		outBaseLayer;
+					};
+				};
+				union
+				{
+					core::vectorSIMDu32 outExtentLayerCount;
+					struct
+					{
+						VkExtent3D		outExtent;
+						uint32_t		outLayerCount;
+					};
+				};
+
+				uint32_t				inMipLevel = 0u;
+				uint32_t				outMipLevel = 0u;
+				ICPUImage* inImage = nullptr;
+				ICPUImage* outImage = nullptr;
+		};
+
+		class CState : public CPrototypeState, public CSummedAreaTableImageFilterBase<ExclusiveMode> {};
```

Comment:
you've just really complicated the inherticance of this class :(

Reply:
corrected in following commit